api/service/syncing/syncing.go

```diff
@@ -327,55 +326,39 @@ func (sc *SyncConfig) GetBlockHashesConsensusAndCleanUp() bool {
 		Int("maxFirstID", maxFirstID).
 		Int("maxCount", maxCount).
 		Msg("[SYNC] block consensus hashes")
-	if float64(maxCount) >= core.ShardingSchedule.ConsensusRatio()*float64(len(sc.peers)) {
-		sc.cleanUpPeers(maxFirstID)
-		return true
-	}
-	return false
+	sc.cleanUpPeers(maxFirstID)
 }
 
 // GetConsensusHashes gets all hashes needed to download.
-func (ss *StateSync) GetConsensusHashes(startHash []byte, size uint32) bool {
-	count := 0
-	for {
-		var wg sync.WaitGroup
-		ss.syncConfig.ForEachPeer(func(peerConfig *SyncPeerConfig) (brk bool) {
-			wg.Add(1)
-			go func() {
-				defer wg.Done()
-				response := peerConfig.client.GetBlockHashes(startHash, size, ss.selfip, ss.selfport)
-				if response == nil {
-					utils.Logger().Warn().
-						Str("peerIP", peerConfig.ip).
-						Str("peerPort", peerConfig.port).
-						Msg("[SYNC] GetConsensusHashes Nil Response")
-					return
-				}
-				if len(response.Payload) > int(size+1) {
-					utils.Logger().Warn().
-						Uint32("requestSize", size).
-						Int("respondSize", len(response.Payload)).
-						Msg("[SYNC] GetConsensusHashes: receive more blockHahses than request!")
-					peerConfig.blockHashes = response.Payload[:size+1]
-				} else {
-					peerConfig.blockHashes = response.Payload
-				}
-			}()
-			return
-		})
-		wg.Wait()
-		if ss.syncConfig.GetBlockHashesConsensusAndCleanUp() {
-			break
-		}
-		if count > TimesToFail {
-			utils.Logger().Info().Msg("[SYNC] GetConsensusHashes: reached retry limit")
-			return false
-		}
-		count++
-		time.Sleep(SleepTimeAfterNonConsensusBlockHashes)
-	}
-	utils.Logger().Info().Msg("[SYNC] Finished getting consensus block hashes")
-	return true
+func (ss *StateSync) GetConsensusHashes(startHash []byte, size uint32) {
+	var wg sync.WaitGroup
+	ss.syncConfig.ForEachPeer(func(peerConfig *SyncPeerConfig) (brk bool) {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			response := peerConfig.client.GetBlockHashes(startHash, size, ss.selfip, ss.selfport)
+			if response == nil {
+				utils.Logger().Warn().
+					Str("peerIP", peerConfig.ip).
+					Str("peerPort", peerConfig.port).
+					Msg("[SYNC] GetConsensusHashes: GetBlockHashes nil response")
+				return
+			}
+			if len(response.Payload) > int(size+1) {
+				utils.Logger().Warn().
+					Uint32("requestSize", size).
+					Int("respondSize", len(response.Payload)).
+					Msg("[SYNC] GetConsensusHashes: receive more blockHahses than request!")
+				peerConfig.blockHashes = response.Payload[:size+1]
+			} else {
+				peerConfig.blockHashes = response.Payload
+			}
+		}()
+		return
+	})
+	wg.Wait()
```

Comment:
Does this goroutine need to block, or can caller move on ?

Reply:
state sync steps are sequential, so this needs to be blocked here.